### PR TITLE
SONARJAVA-912 RSPEC-2445 Synchronized field assignment check

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    ExpressionTree synchronizedExpression = sst.expression();
+    if (isField(synchronizedExpression)) {
+      Symbol field = getSemanticModel().getReference((IdentifierTree) synchronizedExpression);
+      sst.block().accept(new AssignmentVisitor(field, tree));
+    }
+  }
+
+  private boolean isField(ExpressionTree tree) {
+    TypeSymbol enclosingClass = (TypeSymbol) getSemanticModel().getEnclosingClass(tree);
+    if (tree.is(Kind.IDENTIFIER)) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java]: what about : this.myField ?

File: java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    public AssignmentVisitor(Symbol field, Tree tree) {
+      this.field = field;
+      this.synchronizedStatement = tree;
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      AbstractTypedTree variable = (AbstractTypedTree) tree.variable();
+      if (variable.is(Kind.IDENTIFIER)) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java]: Same trouble what about this.myField = "";

File: java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        Symbol variableSymbol = getSemanticModel().getReference((IdentifierTree) variable);
+        if (field.equals(variableSymbol)) {
+          addIssue(synchronizedStatement,
+            MessageFormat.format("Don''t synchronize on \"{0}\" or remove its reassignment on line {1}.", field.getName(), String.valueOf(variable.getLine())));
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java]: No need for String.valueOf

**Comment by benzonico** [comment]
The principle looks good for the current scope of the rule, but what about fields of other classes than the current one : 

synchronize(myObj.myField.MyOtherField) {
   myObj.myField.MyOtherField = "foo"; //should it raise an issue ?
}

**Comment by m-g-sonar** [comment]
Updated to take into accounts reasigned fields of an object field

**Comment by m-g-sonar** [comment]
Updated to correctly handle case of sync on a field of a field.

File: java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    SynchronizedStatementTree sst = (SynchronizedStatementTree) tree;
+    ExpressionTree synchronizedExpression = sst.expression();
+    Symbol field = getField(synchronizedExpression);
+    if (field != null) {
+      sst.block().accept(new AssignmentVisitor(field, tree));
+    }
+  }
+
+  @Nullable
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java]: should be @CheckForNull

File: java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  private Symbol getField(ExpressionTree tree) {
+    if (tree.is(Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) tree;
+      Symbol reference = getSemanticModel().getReference(identifier);
+      if (reference != null && reference.owner().isKind(Symbol.TYP)) {
+        return reference;
+      }
+    } else if (tree.is(Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree mse = (MemberSelectExpressionTree) tree;
+      if (isField(mse.expression())) {
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/SynchronizedFieldAssignmentCheck.java]: Why checking only the expression of the MemberSelectExpression ?

**Comment by benzonico** [comment]
LGTM!

**Pull request closed without merging**